### PR TITLE
[addons] Detect binary addons and Origin Type and set properties to be accessed in skins

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Detect binary addons and Origin Type and set properties to be accessed in skins

## Motivation and Context

Would like to be able to make skin changes to show if an addon is binary an/or a manual install in the AddonBrowser  and AddonInfo screens.

## How Has This Been Tested?

MacOSX

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
@enen92 let me know if the GUIInfoManager entry is correct or not for the string. I didn't see another example to copy.

**Comment by phunkyfish** [comment]
@howie-f if you have time to test the OriginType property on llnux that would be great. Ty

File: xbmc/GUIInfoManager.cpp
Comment (enen92) [inline xbmc/GUIInfoManager.cpp:4395]: You need to escape any commas with a backslash \ otherwise you break the doxygen table

File: xbmc/addons/AddonManager.cpp
Comment (enen92) [inline xbmc/addons/AddonManager.cpp:1091]: Shouldn't this be in addontypes rather than in the main add-on file/definition?

File: xbmc/addons/addoninfo/AddonInfo.h
Comment (enen92) [inline xbmc/addons/addoninfo/AddonInfo.h:40]: Addontypes?

File: xbmc/addons/AddonManager.cpp
Comment (phunkyfish) [reply xbmc/addons/AddonManager.cpp:1091]: It requires the origin so I felt it fit better in the main class. It’s the type of the origin and not the addon itself.

File: xbmc/GUIInfoManager.cpp
Comment (ronie) [inline xbmc/GUIInfoManager.cpp:4397]: typo: ListItem_Property_AddonOrigi**n**Type

File: xbmc/filesystem/AddonsDirectory.cpp
Comment (ronie) [inline xbmc/filesystem/AddonsDirectory.cpp:811]: please do not use hardcoded strings. these need to be localized.

File: xbmc/addons/AddonManager.cpp
Comment (howie-f) [inline xbmc/addons/AddonManager.cpp:1099]: returns:
**repository** for `skin.estouchy`, `resource.images.weatherfanart.single`, `service.xbmc.versioncheck` and more
**system** for `repository.xbmc.org` (which is correct, but it's also a repo)
**manual** for a zip-installed priv. repo

for a quick check i modified a label in DialogAddonInfo.xml:
```
<control type="label">
  <width>470</width>
  <height>40</height>
  <label>$INFO[ListItem.Property(addon.origintype),[COLOR button_focus]$LOCALIZE[31150]:[/COLOR] ]</label>
  <visible>!String.IsEmpty(ListItem.AddonOrigin) + !ListItem.Property(addon.isbinary)</visible>
</control>
```

**Review by howie-f — COMMENTED** [review]
tested on linux:
`addon->IsBinary()` is correctly returning true for `pvr.hts` and `inputstream.adaptive` as well as for `screensaver.xbmc.builtin.black` and others as their `library="dummy.so"`

File: xbmc/filesystem/AddonsDirectory.cpp
Comment (phunkyfish) [reply xbmc/filesystem/AddonsDirectory.cpp:811]: The idea here was to have them enumerated instead of booleans. It should allow logic in the skin such as if Add.OriginType == manual, show this icon. Have I used the wrong approach?

File: xbmc/addons/AddonManager.cpp
Comment (phunkyfish) [reply xbmc/addons/AddonManager.cpp:1099]: Estouchy should come back as system based on the sql output.

Comment (phunkyfish) [reply xbmc/addons/AddonManager.cpp]: Fixed now

**Review by howie-f — APPROVED** [review]
runtime test fine now

File: xbmc/addons/addoninfo/AddonInfoBuilder.cpp
Comment (enen92) [inline xbmc/addons/addoninfo/AddonInfoBuilder.cpp:418]: does this really need to be public? Setting this kind of info should only be done when the addon is parsed so I don't see the need for exposing it

File: xbmc/GUIInfoManager.cpp
Comment (Rechi) [inline xbmc/GUIInfoManager.cpp:4389]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/GUIInfoManager.cpp
Comment (Rechi) [inline xbmc/GUIInfoManager.cpp:4397]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

File: xbmc/addons/addoninfo/AddonInfoBuilder.cpp
Comment (phunkyfish) [reply xbmc/addons/addoninfo/AddonInfoBuilder.cpp:418]: How do you mean public?

File: xbmc/GUIInfoManager.cpp
Comment (phunkyfish) [reply xbmc/GUIInfoManager.cpp:4389]: As far as I’m aware this does not apply to comments. I stand to be corrected however.

File: xbmc/filesystem/AddonsDirectory.cpp
Comment (ronie) [reply xbmc/filesystem/AddonsDirectory.cpp:811]: yes, since this is an infolabel (i.e. it can be displayed as a text label in the skin), those strings need to be added to the language file.
skins can then still use the Addon.OriginType == $LOCALIZE[12345] logic to display an icon.

Comment (phunkyfish) [reply xbmc/filesystem/AddonsDirectory.cpp:811]: Ok, coo, thanks!

Comment (phunkyfish) [reply xbmc/filesystem/AddonsDirectory.cpp:811]: Do I document those localisation IDs somewhere?

File: xbmc/addons/addoninfo/AddonInfoBuilder.cpp
Comment (phunkyfish) [reply xbmc/addons/addoninfo/AddonInfoBuilder.cpp:418]: Removed public function as it's unused.

**Review by enen92 — APPROVED** [review]

File: xbmc/filesystem/AddonsDirectory.cpp
Comment (ronie) [reply xbmc/filesystem/AddonsDirectory.cpp:811]: nope, looks good enough as is to me.

File: xbmc/GUIInfoManager.cpp
Comment (AlwinEsch) [reply xbmc/GUIInfoManager.cpp:4389]: Yes, also can a line break within a doxygen docs make problems see https://github.com/xbmc/xbmc/pull/18174

File: xbmc/addons/addoninfo/AddonInfo.h
Comment (AlwinEsch) [reply xbmc/addons/addoninfo/AddonInfo.h:40]: Think as `Addontypes` can confusing as PVR, Screensaver, Video... also AddonType

**Review by AlwinEsch — APPROVED** [review]
This changes are nice and good to have, thanks a lot!

**Comment by ronie** [comment]
@phunkyfish didn't notice this before, but we already have an `ListItem.AddonOrigin` infolabel.
it provides similar info to the ListItem.Property(Addon.Origintype) one you added.

ListItem.AddonOrigin can return one of the following:
- system
- repository name
- unknown

perhaps the two can/should be merged?


on the subject of ListItem.AddonOrigin, i think one of your recent PR's (https://github.com/xbmc/xbmc/pull/18087) broke this infolabel, as it now always returns 'unknown'. i'm not 100% sure it's your PR, but it's something that got merged around that time.

build from a few days before the PR was merged:
good


build from a few days after the PR was merged:
bad

**Comment by howie-f** [comment]
@ronie despite if the two infolabels should be merged i tested a current master build and get the correct repository for estuary.

screenshot000



could you please verify if your database entries are correct? they should be something like:

```
sqlite> select addonID, origin from installed where addonID like '%estuary%' or addonID like '%hts%';

skin.estuary|b6a50484-93a0-4afb-a01c-8d17e059feda
pvr.hts|

```
`b6a50484-93a0-4afb-a01c-8d17e059feda` is ORIGIN_SYSTEM in core.
for pvr.hts i get an unknown repository (empty origin) because i just copy the files in place.

**Comment by ronie** [comment]
@howie-f thank for looking into it. interesting to see it's working ok at your end.

i did my test with a complete fresh install of todays master.
it updates a few addons at first startup and my addon.db looks to be in good shape:
db

no matter what the 'origin' is, ListItem.AddonOrigin returns 'unknown' for each and every addon.

vc

es

**Comment by ronie** [comment]
this is a test using ListItem.Property(Addon.Origintype) instead. strangely it also returns incorrect results:

es2

it should return 'system' there. estouchy isn't even available in the repo.

i'm starting to wander if this could be some odd platform specific issue? i'm using linux...

**Comment by howie-f** [comment]
yes me too. running on debian. strange, no idea what‘s going on there

**Comment by phunkyfish** [comment]
On mac I see the results I expected. 

As linux was the edge case I did all the testing with Howie to make sure this worked correctly. It appears to not be the case unfortunately.

@howie-f can you somehow debug this?

**Comment by howie-f** [comment]
@phunkyfish will have a look if it started with this pr and then see what i can find

**Comment by ronie** [comment]
i have found the issue, it's due to this change:
https://github.com/xbmc/xbmc/pull/18087/files#diff-130a411077441f19c8eb4b80f5b382d7L168-R169

which inserts a new field into the table.

as a result, you need to update this code as well:
https://github.com/xbmc/xbmc/blob/master/xbmc/addons/AddonDatabase.cpp#L1072-L1076

```
- m_pDS->fv(6).get_asString());
+ m_pDS->fv(7).get_asString());
```

**Comment by ronie** [comment]
and same change needed here as well:
https://github.com/xbmc/xbmc/blob/39f68151963c369a84f57c21a4e894c0409dc421/xbmc/addons/AddonDatabase.cpp#L408

**Comment by howie-f** [comment]
nice find, makes sense as the field was inserted not appended, should i prepare the fix? @phunkyfish let me know and it will happen

**Comment by phunkyfish** [comment]
Go for it @howie-f 😉 Thanks

**Comment by howie-f** [comment]
> and same change needed here as well:
> 
> https://github.com/xbmc/xbmc/blob/39f68151963c369a84f57c21a4e894c0409dc421/xbmc/addons/AddonDatabase.cpp#L408

there seems to be no change needed here as columns are selected explicitely (fieldname repo.addonID is unfortunately misleading, it retrieves origin)

**Comment by ronie** [comment]
> i tested a current master build and get the correct repository for estuary.

> On mac I see the results I expected.

the fact it did work for both of you and not work for me is i was using a freshly created addon.db.
in the case of a fresh addon.db, the 'disabledreason' field is placed **before** the 'origin' field

https://github.com/xbmc/xbmc/blob/39f68151963c369a84f57c21a4e894c0409dc421/xbmc/addons/AddonDatabase.cpp#L167-L169

but with an existing addon.db that gets updated, the 'disabledreason' field is placed **after** the 'origin' field
https://github.com/xbmc/xbmc/blob/39f68151963c369a84f57c21a4e894c0409dc421/xbmc/addons/AddonDatabase.cpp#L223

i'm not an db expert of any kind, so i'm wondering if this inconsistency could potentially bite us in the future?


fresh addon28.db:
db2


addon28.db updated from addon27.db: 
db

**Comment by phunkyfish** [comment]
It should not cause an issue once indexes are not used for accessing results. But it’s probably no harm to move the new field to the end of the create statement too.

**Comment by howie-f** [comment]
it could bite us back if there are more `select * from` - statements in the code and i now only fixed that one causing the issue.
i'd suggest to move `disabledReason` to the end on table creation. or look for `select *` and fix all at once.

**Comment by DaveTBlake** [comment]
It is not  `select * from`  per se that could trip things up, but that in combination with the use of column position in the results dataset rather than field name. Would be good to check for `m_pDS->fv(0)` etc. in AddonDatabase.cpp

**Comment by phunkyfish** [comment]
I think move the table reason to the end on table creation in your PR @howie-f

**Comment by howie-f** [comment]
ok, if no one objects i'll do the bump there and volunteer to search for `fv(0)` etc. at a later point

**Comment by ronie** [comment]
think i have come across another issue...

tested with a fresh install of kodi with https://github.com/xbmc/xbmc/pull/18181 applied:
1) start kodi and install the Southpark addon from our repo
2) next, go to `My Addons > Video addons > Southpark`
the origin is not shown correctly
sp1

3) restart Kodi and go to `My Addons > Video addons > Southpark` again
the origin is now shown correctly
sp2

**Comment by phunkyfish** [comment]
@ronie does this happen with #18181 and @howie-f’s dB fix PR applied?

**Comment by howie-f** [comment]
with #18269 applied and a wiped userfolder before first start i cannot reproduce this now

**Comment by ronie** [comment]
> @ronie does this happen with #18181 and @howie-f’s dB fix PR applied?

sorry, i meant to say current master + https://github.com/xbmc/xbmc/pull/18269

> What about before #18181?

it doesn't seem to be related to any recent changes.
the oldest build i have laying around is from december 2019 and it also happens there.
leia works ok though.

**Comment by phunkyfish** [comment]
Ok, then we need to figure this one out but it shouldn’t hold up howie-f’s fix PR.

I wonder why we can’t repro it.

**Comment by howie-f** [comment]
@phunkyfish @ronie 
it's strange but now i can reproduce the 2nd issue.
at the same time i have to say that #18269 fixes only the first reported, but not the second.
sounds like bisecting to me if dec.2019 was bad and leia was good.... hm. ok needs to be done.
anyway #18269 is save to merge

**Comment by AlwinEsch** [comment]
About them possible that it was come with cpluff kick out (if on Leia OK). Maybe after a fresh install the related part not set correct.
Can look which part relate to there,

**Comment by phunkyfish** [comment]
Thanks @AlwinEsch , hopefully you find the cause!

**Comment by AlwinEsch** [comment]
About this would it be also good to show source if addon not installed?

Bildschirmfoto von 2020-08-15 12-38-04

From my sude yes, as good to know which repo is used for install there.

**Comment by ronie** [comment]
@AlwinEsch there's a PR pending for that
https://github.com/xbmc/xbmc/pull/18266#issuecomment-673173911

**Comment by AlwinEsch** [comment]
Thanks, overseen. About the other thing where after install the origin was missing and shown after restart I thing to found.

Is again a small hen and egg problem. The addon installer set inside database his origin after the addon manager has loaded his new addon infos.

I look now to make this database update within addon manager so that it also direct update the there stored addon infos.

**Comment by howie-f** [comment]
this sounds very promising @AlwinEsch thanks 👍

**Pull request merged**
